### PR TITLE
Shader Graph Automated Node Testing

### PR DESCRIPTION
**Pull request by zadi0648**
### Purpose of this PR
Automated testing for the Shader Graph node library. A document detailing it and how to add more nodes/tests can be found here: https://docs.google.com/document/d/1VmZ071shaBWrFqhcvTzet_Aoij-p-lSRainCH84zrIA/edit#

---
### Testing status
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=tests%2Fsg%2Fnodes&unity_branch=trunk&automation-tools_branch=master

---
### Overall Product Risks
**Technical Risk**: None; automated tests

**Halo Effect**: None; automated tests

### CODE REVIEW

File: TestProjects/ShaderGraph/Assets/CommonAssets/Editor/ShaderGraphParser.cs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.IO;
+using System.Reflection;
+using UnityEngine;
+using UnityEditor;
+using UnityEditor.ShaderGraph;
+using UnityEditor.Graphing;
+
+public class ShaderGraphParser
```
[the remark this answers is not in the document]
Comment (zadi0648) [reply TestProjects/ShaderGraph/Assets/CommonAssets/Editor/ShaderGraphParser.cs]: Okay I could do that, but I leave for a trip next week. In any case this is just a helper script for determining which nodes are in, it doesn't actually get run at all during tests. For the sake of getting the test in could we skip that for now?

[the remark this answers is not in the document]
Comment (pbbastian) [reply TestProjects/ShaderGraph/Assets/CommonAssets/Editor/ShaderGraphParser.cs]: Oh okay, if it doesn't run as part of the test suite then it's ok for now :)

**Comment by sebastienlagarde** [comment]
should we merge this ? or do we wait for something ?

**Comment by pbbastian** [comment]
@zadi0648 @Kink3d Is this ready to go in?

**Comment by zadi0648** [comment]
not yet, we are having an issue with procedural scene

**Comment by zadi0648** [comment]
@pbbastian Long wait is over... That problem we were waiting on finally got resolved, I took care of a few new problems & missing nodes, I also did some cleaning up, test are green in Editor and if Katana tests go green then this is ready to merge.

**Comment by sophiaaar** [comment]
I tried running the OSX tests on Yamato, just for some extra verification, but they will fail without [this](https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/3639) PR I just merged

**Comment by sophiaaar** [comment]
Yamato looks good but I ran the tests in playmode locally on Mac Metal and artistic, input, math, procedural, utility and uv all fail. will show @pbbastian since he is in my time zone

**Comment by sophiaaar** [comment]
Not sure why packing and testing was attempted on this branch - of course it will fail because it validates the packages as if you are going to release them. In this case, it failed because the changelog is incorrect (which is expected at this stage)

**Pull request merged**